### ISSUE TITLE
[Bug?] Tailscale seems to affect DNS even when "Use Tailscale DNS Settings" is disabled

### ISSUE THREAD

**Issue by MahdiNazemi** [tailscale/tailscale#9479]
### What is the issue?

I run Tailscale and Mullvad side-by-side, do not use an exit node in Tailscale, and have the "Use Tailscale DNS Settings" unchecked.

When I turn Tailscale on, Mullvad gets disconnected and shows the following error message:
> Unable to set system DNS server. Please send a problem report.

The same happens when I turn Tailscale off.

### Steps to reproduce

1. Open Tailscale an uncheck the "Use Tailscale DNS Settings" option.

2. Clone Mullvad VPN's source code and change the `mullvadvpn-app/talpid-core/src/firewall/mod.rs` file to bypass the tunnel for traffic to/from 100.64.0.0/10:

```/// When "allow local network" is enabled the app will allow traffic to and from these networks.
pub(crate) static ALLOWED_LAN_NETS: Lazy<[IpNetwork; 7]> = Lazy::new(|| {
    [
        IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(10, 0, 0, 0), 8).unwrap()),
        IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(172, 16, 0, 0), 12).unwrap()),
        IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(192, 168, 0, 0), 16).unwrap()),
        IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(169, 254, 0, 0), 16).unwrap()),
        IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(100, 64, 0, 0), 10).unwrap()),
        IpNetwork::V6(Ipv6Network::new(Ipv6Addr::new(0xfe80, 0, 0, 0, 0, 0, 0, 0), 10).unwrap()),
        IpNetwork::V6(Ipv6Network::new(Ipv6Addr::new(0xfc00, 0, 0, 0, 0, 0, 0, 0), 7).unwrap()),
    ]
});
```

3. Build Mullvad from the source, install the app, enable "Local network sharing", and connect to one of their servers.

4. Enable/Disable Tailscale

### OS

macOS

### OS version

13.5.2

### Tailscale version

1.48.2

### Bug report

BUG-0cc44145f636d76359ba32e22a496df291aee846dad4552f285a4430be9f2ae1-20230920050224Z-2b99cc7d39218941

**Comment by alexl4321** [comment]
can confirm on macos. even if you disable everything like "use tailscale dns", subnets, etc it will still try to manage the dns and break other apps e.g. like a patched mullvad which checks for that.

btw patched mullvad the much better solution than the tailscale mullvad exit partnership...

**Comment by bradfitz** [comment]
Sounds like this a VPN interop problem, not something that Tailscale's doing. If Mullvad says "Unable to set system DNS server. Please send a problem report.", did you send a problem report to Mullvad? I think they meant to send them a problem report, not to us.

**Comment by MahdiNazemi** [comment]
> Sounds like this a VPN interop problem, not something that Tailscale's doing. If Mullvad says "Unable to set system DNS server. Please send a problem report.", did you send a problem report to Mullvad? I think they meant to send them a problem report, not to us.

The response provided lacks substantial contribution to the discussion, merely addressing the superficial aspects of the problem. Prior to posting my bug report, I had already examined the logs of Mullvad and encountered the following line repeatedly:

> [talpid_core::dns::imp][DEBUG] Detected DNS change [[REDACTED],[REDACTED],[REDACTED],[REDACTED]] for State:/Network/Service/[REDACTED]/DNS

Hence, it is evident that the issue stems from Tailscale altering the DNS. Consequently, I find it reasonable to file a bug report here first.

**Comment by bradfitz** [comment]
> Prior to posting my bug report, I had already examined the logs of Mullvad and encountered the following line repeatedly:

You never said that in your original comment.

> Detected DNS change [[REDACTED],[REDACTED],[REDACTED],[REDACTED]] 

So that's evident that it's Tailscale? Is there something in that REDACTED that points to Tailscale?

What DNS change do you see Tailscale making?

Are you sure it's not Mullvad getting confused by a new network interface appearing?

Until there are concrete details about what Tailscale's doing wrong here, there's not much we can do.

It's not a goal of Tailscale to interop alongside all possible VPN programs. Operating systems tend to not like that and not provide suitable APIs for them to even cooperate. (That's why we added Mullvad support to Tailscale directly, to work around their inability to coexist as separate programs)

**Comment by MahdiNazemi** [comment]
@pinkisemils, would you please look at the issue description and let us know your opinion on what is the best course of action to address this problem?

**Comment by alexl4321** [comment]
@MahdiNazemi  hero... works indeed even only after 2nd refresh + you have to do the same thing over again and loose connectivity whenever you switch tailnet accounts. but its a start.

this issue is literally what keeps me from deploying tailscale across multiple orgs with xxx servers.

if i can be of any help let me know!!

**Comment by MahdiNazemi** [comment]
@alexl4321, I'm glad this setup has helped your use cases. I have been using it for a while and find it much better than my previous solution [explained here](https://github.com/tailscale/tailscale/issues/2880#issuecomment-1412829034). The only annoyance is the issue described in this bug report, which, as Brad mentioned, may be an interop issue that's difficult to address.

All my log files show the REDACTED keyword, which prevents a deeper understanding of the problem. If you know how to print the full logs without redacting any information, it can help us pinpoint the actual problem.

**Comment by alexl4321** [comment]
@MahdiNazemi yeah i did similar solution first with the official wireguard but having mullvad-app is much better for a variety of reasons.

still having some issues now and then and latest mullvad client master has a bug but kinda working with 2023.4.

what i am doing now is running the 100.x patched mullvad client and adding a static route for 100.x on macos to a raspberry pi in the network which runs tailscale and let that NAT all traffic for that subnet into tailscale. that works lovely without any of the hiccups and nothing on the mac machine interferes with the mullvad.

only problem i still have there is i need to access multiple tailnets so what i will try is to run the 3 tailnets in docker containers on the mac itself and then add routes to them depending on the target tailnet. kinda big list though since tailnet does not have fixed subnets per tailnet so its all over the place :( but should do the final trick until hopefully this all works better together. could share all tailnets into one of course but also not really a solution.

**Comment by fgionghi** [comment]
I have the same problem using tailscale and netbird. 
I have unchecked DNS settings on tailscale and using netbird to resolve my.domain.

When I turn tailscale up and do a `host x.my.domain` it gives me:
```
;; communications error to 127.0.0.53#53: timed out
;; communications error to 127.0.0.53#53: timed out
;; no servers could be reached
```
using `tcpdump -i any port 53 -nnn | grep x.my.domain` the difference I see between having tailscale up or down are these:
With tailscale up:
`11:10:51.917093 lo    In  IP 100.78.38.48.57043 > 100.78.38.48.53: 22391+ A? x.my.domain. (36)`
So it is trying to correctly ask to netbird on port 53 to resolve x.my.domain, but seems not responding.

If I turn tailscale down I get:
`
11:13:30.528160 lo    In  IP 100.78.38.48.55229 > 100.78.38.48.53: 42184+ A? x.my.domain. (36)
11:13:30.528475 wt0   Out IP 100.78.38.48.55632 > my-dns-server-ipaddress: 42184+ A? x.my.domain. (36)
`
now is working.
If you need other logs let me know.

**Comment by alexl4321** [comment]
It is the tailscale gui app. If you install tailscale cli via brew it does not have this problem btw...

**Comment by fgionghi** [comment]
I'm on ubuntu 22.04, don't have the gui and didn't know there is one.
I'm using tailscale 1.62.0.

**Comment by virtuous-byte** [comment]
@MahdiNazemi are you still using this setup? I just did all the steps to make Mullvad work along with Tailscale, and even with Tailscale DNS on I'm now able to my tailnet and still access web through Mullvad, but somehow my Tailscale speed is pretty low compared to direct LAN access and Mullvad being off. For example I can get around 200-300Mbps with LAN at 1ms latency, and a similar result with Tailscale on and Mullvad off, but as soon as I connect to Mullvad, Tailscale speed drops to 20~ Mbps. I used iperf3 to calculate those speeds, and even during the test the speed reduces to just 0 b/s in some parts, then continues at 20~ again. Did you have a similar experience while both are on? Theorically I can use Tailscale now, but realistically it's unusably slow.
